src/content/en/fundamentals/engage-and-retain/push-notifications/web-push-protocol.md

```diff
@@ -0,0 +1,613 @@
+attacks as it should only be used once.
+
+The content encryption key (CEK) is the key that will ultimately be used to encrypt our payload.
+
+First we need to create the bytes of data for the nonce and CEK, which is simply a content
+encoding string followed by the context buffer we just calculated:
+
+    const nonceEncBuffer = new Buffer('Content-Encoding: nonce\0', 'utf8');
+    const nonceInfo = Buffer.concat([nonceEncBuffer, contextBuffer]);
+
+    const cekEncBuffer = new Buffer('Content-Encoding: aesgcm\0');
+    const cekInfo = Buffer.concat([cekEncBuffer, contextBuffer]);
+    
+This information is run through HKDF combining the salt and PRK with the nonceInfo and cekInfo:
+
+    // The nonce should be 12 bytes long
+    const nonce = hkdf(salt, prk, nonceInfo, 12);
+
+    // The CEK should be 16 bytes long
+    const contentEncryptionKey = hkdf(salt, prk, cekInfo, 16);
+    
+This gives us our nonce and content encryption key.
+
+### Perform the Encryption
+
+Now that we have our content encryption key, we can encrypt the payload.
+
+We create an AES128 Cipher using the content encryption key
+as the key and the nonce is as an initialization vector.
+
+In Node this is done like so:
+
+    const cipher = crypto.createCipheriv(
+      'id-aes128-GCM', contentEncryptionKey, nonce);
+    
+Before we encrypt our payload, we need to define how much padding we wish to
+add to the front of the payload. The reason you'd want to add padding is that
+it further hides what the content of the payload *could* be.
+
+The padding needs to be at least 2 bytes and these first two bytes should indicate the length
+of the padding. If you do add padding, create enough bytes of data to add the two bytes used to
+include the padding length.
+
+    const padding = new Buffer(2 + paddingLength);
+    // The buffer must be only zeros, except the length
+    padding.fill(0);
+    padding.writeUInt16BE(paddingLength, 0);
+    
+We then run our padding and payload through this cipher.
+
+    const result = cipher.update(Buffer.concat(padding, payload));
+    cipher.final();
+
+    // Append the auth tag to the result -
+    // https://nodejs.org/api/crypto.html#crypto_cipher_getauthtag
+    const encryptedPayload = Buffer.concat([result, cipher.getAuthTag()]);
+    
+We now have our encrypted payload - Yay.
+
+The remaining piece of work is to determine how this payload is sent to the push service.
+
+### Encrypted Payload Headers & Body
+
+To send this encrypted payload to the Push Service we need to define a few
+different headers in our POST request.
+
+#### Encryption Header
+
+The 'Encryption' header must contain the *salt* used for encrypting the payload.
+
+The 16 byte salt should be base64 URL safe encoded and added to the Encryption header, like so:
+
+    Encryption: salt=<URL Safe Base64 Encoded Salt>
+    
+#### Crypto-Key Header
+
+We saw that the 'Crypto-Key' header is used under the 'Application Server Keys'
```

Comment:
I think it's `Crypto-Key` not 'Crypto-Key'. Please make this change to all header names throughout.